### PR TITLE
Standalone immutable docids

### LINKED ISSUE #1272: Make DocIDs immutable, set flag for delete, dont serve flagged entries at search time
## Current State

In the current implementation of the inverted index, documents are mutable. As a result an update might remove some inverted-index properties and add others. However, we have seen so far that this is problematic:

* The HNSW vector index is immutable by design, so we are in a messy situation where we currently don't assign a new ID when the vector is unchanged, but do assign a new ID if the vector is not changed.
* The above becomes even more problematic when specific properties have property-specific indices, such as with the Geo-Index, which in turn are immutable, so an update would also require a new doc ID for this
* Inverted-Index Deletes are very slow at scale (1M+ entries in the db) as our scale tests with the arXiv dataset have shown if an update happens at this scale, deleting from an inverted index row is very slow. It would be much more efficient if we could just stay in "append-only" mode.

## Goals

* [x] DocIDs (and therefore all entries in the inverted and vector indices) are immutable
* [x] On update, we assign a new one and mark the old one as deleted (very cheap)
  * ~~this will probably have to be held in mem and be written to disk~~
  * ~~possible shortcut could include to remove the in mem cache part, however, this might produce considerably worse performace as every single read would have to check the "what is currently deleted" store from disk~~
  * as part of the technical analysis, we have found out that we can simply use the existing `docID->UUID` lookup table which is now extended to also contain deletion status and deletion timestamp
* [x] At search time, we check the deletion list / tombstone if an id we encountered has already been deleted, if yes, it is ignored. This means a delete is consistent to the user as a deleted entry can no longer be served, once a tombstone is in.
* [x] complex decision logic about when to increase the doc id and when to mutate it, is removed. (This should lead to better code as well)

## Out of scope
* the periodic removal has been scoped out to #1285 

## Process
This issue should probably be coordinated with the implementation of #1282, they will probably touch on very similar areas. So if they can happen in sequence as opposed to parallel, we will probably avoid plenty of merge conflicts.

### LINKED ISSUE #1281: Standalone support updates on geo props
**Blocked** - depends on completion of #1249 and #1272 first

## Background 

In #1249 we could only support importing geo Props, but not updating them due to the issues outlined in #1272. This issues assumes that #1272 has been fixed and completes geo Props

## Goals

* [x] an object containing a geo prop can be updated
* [x] the old docID is removed from the property-specific geo Index
* [x] the new one is added
* [x] searches correctly use the new geo position and no longer the old one
* [x] support batch updates

### PR DESCRIPTION
**Pull request by etiennedi**
* closes #1272 (make doc ids immutable)
* closes #1281 (support geo index updates)
* makes doc ids immutable, the inverted index is now write only
* anything marked as deleted is not served on an incoming search request
* deleted doc ids are not currently cleaned up from the inverted index (will be done async in #1285)
* contains refactors
  * consistently call doc id `docID` (it used to be sometimes "index id", etc.)
  * split out anything related to `docid` into separate package
  * refactor any method that's too long I encountered on the way

### CODE REVIEW

File: adapters/repos/db/shard_docid.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2020 SeMI Technologies B.V. All rights reserved.
+//
+//  CONTACT: hello@semi.technology
+//
+
+package db
```
Comment (antas-marcin) [inline adapters/repos/db/shard_docid.go:12]: I think that this file should be deleted

File: adapters/repos/db/shard_geo_props.go [hunk 2]
```diff
@@ -93,3 +93,73 @@ func (s *Shard) makeCoordinatesForID(propName string) geo.CoordinatesForID {
...
+	if err := index.GeoIndex.Add(int(status.docID), asGeo); err != nil {
+		return errors.Wrapf(err, "insert into geo index")
+	}
+
+	return nil
+}
+
+func (s *Shard) deleteFromGeoIndex(index propertyspecific.Index,
+	docID uint32) error {
+	fmt.Printf("delete doc id %d\n", docID)
```
Comment (antas-marcin) [inline adapters/repos/db/shard_geo_props.go:159]: I think this line should be also deleted

Comment (etiennedi) [reply adapters/repos/db/shard_geo_props.go:159]: :D Probably a leftover from furious `Prinft` debugging -)

File: adapters/repos/db/vector/hnsw/delete.go [hunk 3]
```diff
@@ -23,9 +23,62 @@ import (
...
+}
+
+func (h *hnsw) tombstonesAsDenyList() helpers.AllowList {
+	deleteList := helpers.AllowList{}
+	h.RLock()
+	tombstones := h.tombstones
+	for id := range tombstones {
+		deleteList.Insert(uint32(id))
+	}
+	h.RUnlock()
```
Comment (antas-marcin) [inline adapters/repos/db/vector/hnsw/delete.go:77]: I would change that to `defer h.RUnlock`

File: adapters/repos/db/vector/hnsw/delete.go [hunk 4]
```diff
@@ -23,9 +23,62 @@ import (
...
+	// This is especially true if the tombstoned entrypoint is the only node in
+	// the graph. In this case we must reset the graph, so it acts like an empty
+	// one. Otherwise we'd insert the next id and have only one possible node to
+	// connect it to (the entrypoint). With that one being tombstoned, the new
+	// node would be guaranteed to have zero edges
+	denyList := h.tombstonesAsDenyList()
+
+	h.RLock()
+	node := h.nodes[id]
+	h.RUnlock()
```
Comment (antas-marcin) [inline adapters/repos/db/vector/hnsw/delete.go:42]: I would make this a separate method just for getting the given node by id

**Pull request merged**